### DISCUSSION TITLE
How should I be starting an ebiten mobile project?

### DISCUSSION THREAD

**Discussion by airtonix** [hajimehoshi/ebiten#2172] [Q&A]
Inspired by the [ebitenmobile document](https://ebiten.org/documents/mobile.html), i set out to turn a simple desktop playground into adesktop/mobile monorepo.

Disclaimer, I've only just learnt golang in the last few days.

# 1. First Attempt

Starting with the [ebiten-aseprite/example](https://github.com/SolarLune/goaseprite/tree/ab8a34ca84583951e4a16987bcf248e9fa8cc0b2/example) directory, I attempted to reproduce it running on my android phone.

1. add the `mobile.go` as described by the blog post.
2. run the `ebitenmobile bind ...` command

😢  No success

- errors about `ebitenutil.NewImageFromFile` being undefined or something?
- running and installing the example from https://github.com/golang/go/wiki/Mobile works 
- running `gomobile install -tags=example github.com/hajimehoshi/ebiten/v2/examples/paint` works

# 2. Second Attempt

So I thought maybe the errors about `ebitenutil.NewImageFromFile` being undefined are something to do with a filesystem problem.

So I looked at what https://github.com/hajimehoshi/go-inovation was doing with regards to loading sprites. 

I noticed it was there was some relationship between something called `embed`, some magic comments and how they were accessed.

I set out to replicate that. 

But first I split the repo into more of a monorepo structure.

```sh
cmd/
  desktop/
    main.go
pkg/
  game/
    assets/
      images/
        *.png/json/aseprite
    core/
      aseprite.go
      images.go
    game.go    
```

I added the mobile cmd according to the [ebitenmobile document](https://ebiten.org/documents/mobile.html)

```sh
cmd/
  desktop/
    main.go
  mobile/
    mobile.go
pkg/
  game/
    assets/
      assets.go 🗨️ 
      images/
        *.png/json/aseprite
    core/
      aseprite.go
      images.go
    game.go 
```

result: https://github.com/airtonix/golang-ebiten-multiplatform-aseprite-monorepo/tree/1491c44

😢  No success

- running `gomobile install ./cmd/mobile` errors, complaining it can't find an apk

# 3. Third Attempt

So from my experience that is react-native on ios and android, my next thought was that I need to provide all the ugly and boring files that android studio and xcode like to magically generate 🤮 .

So i proceeded to copy what ever was in the `android` dir from https://github.com/hajimehoshi/go-inovation, replacing any app id references in file contents or file/folder names

I end up with 

```sh
cmd/
  desktop/
    main.go
  mobile/
    android/
      🤮 *
    mobile.go
pkg/
  game/
    assets/
      assets.go 🗨️ 
      images/
        *.png/json/aseprite
    core/
      aseprite.go
      images.go
    game.go 
```

- running `gomobile install ./cmd/mobile` works and now I have an app on my phone... but it just hangs when i launch it and have to force kill it.

# Conclusion

So this is where I'm at. 

I'll be coming back to this to understand more about what's required.

However, it might be worth updating the [ebitenmobile document](https://ebiten.org/documents/mobile.html) to mention that nothing will work until the user provides the `android` and `ios` project files generated by respective android and ios tools.

My other thought was to provide a generator that could make use of https://github.com/StephanHCB/go-generator-cli/ to quickly scaffold these two directories.

**Comment by hajimehoshi** [comment]
This sounds like how to use go:embed. If go:embed succeeds, you no longer have to care about the location of your resources, right? I'm not sure what failed in your second attempt.

**Reply by airtonix** [reply]
apologies, perhaps the title misrepresented that I was trying to ask a question :)

**Comment by staatzstreich** [comment]
Hi, i am also interested in this topic. I have created a simple card game that's running on Mac and WebAssembly. All work fine so far. Now I want to try doing an IOS Build but struggling with warnings and error messages.
I grabbed a copy of go-inovation and executed the steps written at the documentation .... same result. Warnings and errors as follow:

```
gomobile: ios/arm64: go build -buildmode=c-archive -o /var/folders/87/bjbghrss6wx045w885y2xqd40000gn/T/gomobile-work-4219530336/Mobile-ios-arm64.a ./gobind failed: exit status 2
# golang.org/x/mobile/app
darwin_ios.m:89:74: warning: 'statusBarOrientation' is deprecated: first deprecated in iOS 13.0 - Use the interfaceOrientation property of the window scene instead. [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:146:55: note: property 'statusBarOrientation' is declared deprecated here
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:146:55: note: 'statusBarOrientation' has been explicitly marked deprecated here
darwin_ios.m:97:75: warning: 'statusBarOrientation' is deprecated: first deprecated in iOS 13.0 - Use the interfaceOrientation property of the window scene instead. [-Wdeprecated-declarations]
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:146:55: note: property 'statusBarOrientation' is declared deprecated here
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/System/Library/Frameworks/UIKit.framework/Headers/UIApplication.h:146:55: note: 'statusBarOrientation' has been explicitly marked deprecated here
# gobind/gobind
mobileebitenviewcontroller_ios.m:10:9: fatal error: 'GLKit/GLkit.h' file not found
mobileebitenviewcontroller_ios.m:10:9: note: did not find header 'GLkit.h' in framework 'GLKit' (loaded from '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS15.5.sdk/System/Library/Frameworks')

exit status 1 
```

GLKit.framework is present at mentioned path..... when looking into the "GLKit.framework" folder ... the header file is there.. with using the parameter "-iosversion 12.0" the notes about 'statusBarOrientation'  can be suppressed....

I am not skilled with building IOS Applications nor Xcode etc... have no clue what to do now.... Any Tips are welcome 

thx Michael

**Reply by hajimehoshi** [reply]
> mobileebitenviewcontroller_ios.m:10:9: fatal error: 'GLKit/GLkit.h' file not found

Hmm, interesting. Let me investigate.

The warning should not be a blocker by the way.

**Reply by hajimehoshi** [reply]
I think I could fix the bug. Could you try 7fdbb4a5d5d6d47623578743e18d25ef0959ecb2? In your project, run these commands:

```
go get github.com/hajimehoshi/ebiten/v2@7fdbb4a5d5d6d47623578743e18d25ef0959ecb2
go install github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile
```

**Reply by staatzstreich** [reply]
Hi ... thank you for that fast reply!! Compilation now works, your fix was the right one... but in the end there is now the message `"open mobile/ios/patience.xcframework/ios-arm64_x86_64-simulator/patience.framework/Versions/A/Headers/MobileEbitenViewController.h: no such file or directory" 
`

**Reply by staatzstreich** [reply]
As far as I understand the whole thing .... the path "....simulator/patience.framework/..." is not found because it is written as "...simulator/Patience.framework/..." on my disk and I am using a case sensitive formatted disk.... something in the build process seems to write upper case ...

**Reply by hajimehoshi** [reply]
Thanks, I'll take a look further.

**Reply by hajimehoshi** [reply]
When was the error message shown? During the execution of ebitenmobile?

**Reply by staatzstreich** [reply]
Yes

**Reply by hajimehoshi** [reply]
OK, could you try 4ba0ab6e3a90de24444025481d08f8d8ab595620? Thanks,

```
go get github.com/hajimehoshi/ebiten/v2@4ba0ab6e3a90de24444025481d08f8d8ab595620
go install github.com/hajimehoshi/ebiten/v2/cmd/ebitenmobile
```

**Reply by staatzstreich** [reply]
Thx a lot for helping.... "ebitenmobile bind" is now finishing without any errors .... 👍

**Reply by hajimehoshi** [reply]
Thank you for confirming the fix!

**Reply by staatzstreich** [reply]
Hi, me again...

After successfully create a "binding" I have tried to create an IOS App. First Problem for me was that the Simulator would not start because of missing a Info.plist file...
I could fix this with copying the file from ".xcframework" Folder into the both subfolders fx. "iOS-arm64" and "ios-arm64_x86_64-simulator"....

I am am not sure doing this right.... The documentation is different from the mentioned "go-inovation" project....
I have tried several things and most resulted in a white screen or an error message that the Interface Builder can't find the ViewController.

My last attempt as following:

Using Mac Monterey 12.4 and Xcode Version 13.4.1 (13F100):
1.) start a new Project => iOS => App => Storyboard + Objective-C
2.) Added Framework at the "General" Tab of Target Settings chooses the generated ".xcframework" folder and selected "Do Not Embed"
3.) Modified the ViewController.h to:
```
#import <UIKit/UIKit.h>
#import "Patience/MobileEbitenViewController.h"

@interface ViewController : MobileEbitenViewController


@end 
```

4.) and the ViewController.m is unedited:
```
#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}


@end
```

At Storyboard the ViewController has "Is Initial View Controller" checked at the Attributes Inspector

building the project resulted in following error messages:

```
Ld /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Products/Debug-iphonesimulator/Patience.app/Patience normal (in target 'Patience' from project 'Patience')
    cd /Users/michaelstaatz/CodeProjects/patience-max/mobile/ios/Patience
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-ios15.5-simulator -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator15.5.sdk -L/Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Products/Debug-iphonesimulator -F/Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Products/Debug-iphonesimulator -filelist /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Intermediates.noindex/Patience.build/Debug-iphonesimulator/Patience.build/Objects-normal/x86_64/Patience.LinkFileList -Xlinker -rpath -Xlinker @executable_path/Frameworks -dead_strip -Xlinker -object_path_lto -Xlinker /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Intermediates.noindex/Patience.build/Debug-iphonesimulator/Patience.build/Objects-normal/x86_64/Patience_lto.o -Xlinker -export_dynamic -Xlinker -no_deduplicate -Xlinker -objc_abi_version -Xlinker 2 -fobjc-arc -fobjc-link-runtime -Xlinker -sectcreate -Xlinker __TEXT -Xlinker __entitlements -Xlinker /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Intermediates.noindex/Patience.build/Debug-iphonesimulator/Patience.build/Patience.app-Simulated.xcent -framework Patience -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Intermediates.noindex/Patience.build/Debug-iphonesimulator/Patience.build/Objects-normal/x86_64/Patience_dependency_info.dat -o /Users/michaelstaatz/Library/Developer/Xcode/DerivedData/Patience-aipwjfbqxltvpuedwjchmqviovus/Build/Products/Debug-iphonesimulator/Patience.app/Patience

Undefined symbols for architecture x86_64:
  "_GCControllerDidConnectNotification", referenced from:
      _initializeGamepads in Patience(000039.o)
  "_GCControllerDidDisconnectNotification", referenced from:
      _initializeGamepads in Patience(000039.o)
  "_GCInputDualShockTouchpadButton", referenced from:
      _getControllerState in Patience(000039.o)
      _getControllerPropertyFromController in Patience(000039.o)
  "_GCInputXboxPaddleOne", referenced from:
      _getControllerState in Patience(000039.o)
      _getControllerPropertyFromController in Patience(000039.o)
  "_GCInputXboxPaddleTwo", referenced from:
      _getControllerState in Patience(000039.o)
      _getControllerPropertyFromController in Patience(000039.o)
  "_GCInputXboxPaddleThree", referenced from:
      _getControllerState in Patience(000039.o)
      _getControllerPropertyFromController in Patience(000039.o)
  "_OBJC_CLASS_$_GCController", referenced from:
      objc-class-ref in Patience(000039.o)
  "_GCInputXboxPaddleFour", referenced from:
      _getControllerState in Patience(000039.o)
      _getControllerPropertyFromController in Patience(000039.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Is this because I try to use the Simulator? Are there things I am doing wrong? 

so far .... as I said... I have no experience with Xcode and Objective-C ... I have only a little experience with Swift and Swift-UI since I have coded some examples from "Hacking with Swift" ....

best,
Michael

**Reply by hajimehoshi** [reply]
Could you file a new issue for that? As I'm not familiar with iOS simulators, I'm not sure I can fix this.

**Reply by staatzstreich** [reply]
Okay... i will do so... Maybe i will also create an discord account and ask there too.
